features/events/api/src/main/java/org/opennms/netmgt/events/api/model/IAlarmData.java

```diff
@@ -35,11 +35,14 @@
  */
 public interface IAlarmData {
     Integer getAlarmType();
+    Integer copyAlarmType();
```

Comment:
As discussed lets see if we can get rid of these `copy...()` methods by updating the `copyFrom()` methods to interrogate the object via the `has...()` methods to determine if a default would be given or not.

Reply:
If the ImmutableMapper implementations are updated to check the `has` method first and set the appropriate value on the Immutable object we should be able to get rid of the copy methods on the interface since we know the value returned by the getter will be as-set not a default.

Reply:
Ex using `AlarmData`:

```
    public static ImmutableAlarmData fromMutableAlarmData(AlarmData alarmData) {
        if (alarmData == null) {
            return null;
        }

        return ImmutableAlarmData.newBuilder()
                .setReductionKey(alarmData.getReductionKey())
                .setAlarmType(alarmData.hasAlarmType() ? alarmData.copyAlarmType() : null)
                .setClearKey(alarmData.getClearKey())
                .setAutoClean(alarmData.hasAutoClean() ? alarmData.getAutoClean() : null)
                .setX733AlarmType(alarmData.getX733AlarmType())
                .setX733ProbableCause(alarmData.hasX733ProbableCause() ? alarmData.copyX733ProbableCause() : null)
                .setUpdateFieldList(
                        alarmData.getUpdateFieldList()
                                .stream().map(ImmutableMapper::fromMutableUpdateField)
                                .collect(Collectors.toList()))
                .setManagedObject(fromMutableManagedObject(alarmData.getManagedObject()))
                .build();
    }
```

Reply:
Updated to remove the `copy...()` methods and make use of `has...()` methods.  Thanks for the feedback.  :)